packages/webhooks/tasks/sendWebhook.js

```diff
@@ -0,0 +1,144 @@
+            // Step 1: Receive another batch
+            const objs = await execGqlAsUser(keystone, user, {
+                query,
+                variables: {
+                    first: maxPackSize || DEFAULT_MAX_PACK_SIZE,
+                    skip: totalLoaded,
+                    where,
+                },
+                dataPath: 'objs',
+                deleted: true,
+            })
+
+            lastLoaded = objs.length
+            // time is measured by the time of the last response from our server received
+            lastSyncTime = dayjs().toISOString()
+
+            // No more objects -> GOTO final update syncedAt
+            if (lastLoaded === 0) {
+                break
+            }
+            // Step 2: Send batch to specified url
+            const response = await trySendData(url, objs)
+            lastSendSuccess = response.ok
+
+            // If not succeed - log and finish task
+            if (!lastSendSuccess) {
+                const noFailureIncrementInterval = dayjs().subtract(1, 'hour')
+                const lastUpdate = dayjs(lastSubscriptionUpdate)
+                // NOTE: If no failures before or > 1 hour since last failure passed -> increment failuresCount
+                if (failuresCount === 0 || noFailureIncrementInterval.isAfter(lastUpdate)) {
+                    await WebhookSubscription.update(keystone, subscriptionId, {
+                        failuresCount: failuresCount + 1,
+                        dv: 1,
+                        sender: { dv: 1, fingerprint: 'sendWebhook' },
+                    })
+                }
+
+                if (response.status === 523) {
+                    logger.error({ message: 'Data was not sent. Reason: Unreachable resource', subscriptionId })
+                    return { status: NO_RESPONSE_STATUS }
+                } else {
+                    logger.error({ message: 'Data was sent, but server response was not ok', status: response.status, subscriptionId })
+                    return { status: BAD_RESPONSE_STATUS }
+                }
+
+            } else {
+                totalLoaded += lastLoaded
+                logger.info({ message: 'Data batch was sent', subscriptionId, batchSize: lastLoaded, total: totalLoaded, syncedAt })
+            }
+
+            // Step 2: Update subscription state if full batch received. Else condition will lead to final update
+            if (lastLoaded === packSize) {
+                await WebhookSubscription.update(keystone, subscriptionId, {
+                    syncedAmount: totalLoaded,
+                    dv: 1,
+                    sender: { dv: 1, fingerprint: 'sendWebhook' },
+                })
+            }
+        }
+
+        await WebhookSubscription.update(keystone, subscriptionId, {
+            syncedAt: lastSyncTime,
+            dv: 1,
+            sender: { dv: 1, fingerprint: 'sendWebhook' },
+        })
+
+        return { status: OK_STATUS }
+
+    } finally {
```

Comment:
Is here any error when webhook is at "processing status" and lock wasn't released yet?

Reply:
As far as I know....if 30 sec will pass, and hook is not finished, then it will throw an error. Also fetch (before I wrapped it) was throwable...So I guess this construction is safe enough